File: internal/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"errors"
	"fmt"
)

const (
	// Version1 is the initial wire-format version.
	Version1 byte = 1

	// CurrentVersion is the protocol version written by Encode.
	CurrentVersion = Version1

	// HeaderSize is the number of bytes used for the version+command-ID
	// header (1B version + 4B big-endian commandID).
	// Exported so transport layers can compute connection-level read limits
	// without duplicating the value.
	HeaderSize = 5

	// MaxPayloadSize is the maximum allowed payload size in bytes (10 MB).
	// Exported so transport layers can enforce this limit at the wire level
	// before any protocol buffer is allocated.
	MaxPayloadSize = 10 * 1024 * 1024
)

// ErrUnsupportedVersion is returned by Decode when the frame's version byte
// does not match a version this build understands.
var ErrUnsupportedVersion = errors.New("unsupported protocol version")

// Encode encodes a frame as: [1B version][4B BE commandID][payload].
// The version is always CurrentVersion.
func Encode(commandID uint32, payload []byte) ([]byte, error) {
	if len(payload) > MaxPayloadSize {
		return nil, fmt.Errorf("payload size %d exceeds maximum %d bytes", len(payload), MaxPayloadSize)
	}

	out := make([]byte, HeaderSize+len(payload))
	out[0] = CurrentVersion
	binary.BigEndian.PutUint32(out[1:HeaderSize], commandID)
	copy(out[HeaderSize:], payload)
	return out, nil
}

// Decode decodes a frame's version byte, commandID (big-endian) and payload
// (returned as an independent copy of the input slice).
//
// Returns ErrUnsupportedVersion if the version byte is not CurrentVersion, so
// callers can distinguish a version mismatch from a malformed frame.
//
// Returning a payload copy guarantees that:
//   - The caller may safely retain or modify the payload without affecting the
//     original network buffer.
//   - Concurrent handler goroutines receive distinct memory and cannot
//     accidentally corrupt each other's data.
func Decode(data []byte) (version byte, commandID uint32, payload []byte, err error) {
	if len(data) < HeaderSize {
		return 0, 0, nil, errors.New("data too short")
	}

	version = data[0]
	if version != CurrentVersion {
		return version, 0, nil, ErrUnsupportedVersion
	}

	payloadSize := len(data) - HeaderSize
	if payloadSize > MaxPayloadSize {
		return version, 0, nil, fmt.Errorf("payload size %d exceeds maximum %d bytes", payloadSize, MaxPayloadSize)
	}

	commandID = binary.BigEndian.Uint32(data[1:HeaderSize])
	payload = make([]byte, payloadSize)
	copy(payload, data[HeaderSize:])
	return version, commandID, payload, nil
}

File: clients/js/src/protocol.ts
// Wire format: [1B version][4B BE commandID][payload]
// Mirrors internal/protocol/protocol.go.

export const CURRENT_VERSION = 1;
export const HEADER_SIZE = 5;
export const MAX_PAYLOAD_SIZE = 10 * 1024 * 1024;

export const ReservedCommands = {
  JsonRpc: 0xffffffff,
  JsonRpcError: 0xfffffffe,
  Ping: 0xfffffffd,
} as const;

export class UnsupportedVersionError extends Error {
  readonly version: number;
  constructor(version: number) {
    super(`unsupported protocol version: ${version}`);
    this.version = version;
  }
}

export function encodeFrame(commandId: number, payload: Uint8Array): Uint8Array {
  if (payload.length > MAX_PAYLOAD_SIZE) {
    throw new Error(`payload size ${payload.length} exceeds maximum ${MAX_PAYLOAD_SIZE} bytes`);
  }
  const out = new Uint8Array(HEADER_SIZE + payload.length);
  const view = new DataView(out.buffer);
  view.setUint8(0, CURRENT_VERSION);
  view.setUint32(1, commandId, false);
  out.set(payload, HEADER_SIZE);
  return out;
}

export interface DecodedFrame {
  version: number;
  commandId: number;
  payload: Uint8Array;
}

export function decodeFrame(data: ArrayBufferLike): DecodedFrame {
  if (data.byteLength < HEADER_SIZE) {
    throw new Error("data too short");
  }
  const view = new DataView(data);
  const version = view.getUint8(0);
  if (version !== CURRENT_VERSION) {
    throw new UnsupportedVersionError(version);
  }
  const commandId = view.getUint32(1, false);
  const payload = new Uint8Array(data.slice(HEADER_SIZE));
  return { version, commandId, payload };
}
